receiver/opencensusreceiver/octrace/opencensus.go

```diff
@@ -146,63 +105,68 @@ func (ocr *Receiver) Export(tes agenttracepb.TraceService_ExportServer) error {
 	}
 }
 
-// Stop the receiver and its workers
-func (ocr *Receiver) Stop() {
-	for _, worker := range ocr.workers {
-		worker.stopListening()
+func (ocr *Receiver) processReceivedMsg(
+	ctx context.Context,
+	lastNonNilNode *commonpb.Node,
+	resource *resourcepb.Resource,
+	recv *agenttracepb.ExportTraceServiceRequest,
+) (*commonpb.Node, *resourcepb.Resource, error) {
+	// If a Node has been sent from downstream, save and use it.
+	if recv.Node != nil {
+		lastNonNilNode = recv.Node
 	}
-}
 
-type receiverWorker struct {
-	receiver *Receiver
-	cancel   chan struct{}
-}
-
-func newReceiverWorker(receiver *Receiver) *receiverWorker {
-	return &receiverWorker{
-		receiver: receiver,
-		cancel:   make(chan struct{}),
+	// TODO(songya): differentiate between unset and nil resource. See
+	// https://github.com/census-instrumentation/opencensus-proto/issues/146.
+	if recv.Resource != nil {
+		resource = recv.Resource
 	}
-}
 
-func (rw *receiverWorker) listenOn(cn <-chan *traceDataWithCtx) {
-	for {
-		select {
-		case tdWithCtx := <-cn:
-			rw.export(tdWithCtx.ctx, tdWithCtx.data)
-		case <-rw.cancel:
-			return
-		}
+	td := &consumerdata.TraceData{
+		Node:         lastNonNilNode,
+		Resource:     resource,
+		Spans:        recv.Spans,
+		SourceFormat: "oc_trace",
 	}
-}
 
-func (rw *receiverWorker) stopListening() {
-	close(rw.cancel)
+	err := ocr.sendToNextConsumer(ctx, td)
+	return lastNonNilNode, resource, err
 }
 
-func (rw *receiverWorker) export(longLivedCtx context.Context, tracedata *consumerdata.TraceData) {
+func (ocr *Receiver) sendToNextConsumer(longLivedCtx context.Context, tracedata *consumerdata.TraceData) error {
 	if tracedata == nil {
-		return
+		return nil
 	}
 
 	if len(tracedata.Spans) == 0 {
-		return
+		observability.RecordMetricsForTraceReceiver(longLivedCtx, 0, 0)
+		return nil
 	}
 
 	// Trace this method
 	ctx, span := trace.StartSpan(context.Background(), "OpenCensusTraceReceiver.Export")
 	defer span.End()
 
-	// TODO: (@odeke-em) investigate if it is necessary
-	// to group nodes with their respective spans during
-	// spansAndNode list unfurling then send spans grouped per node
-
 	// If the starting RPC has a parent span, then add it as a parent link.
 	observability.SetParentLink(longLivedCtx, span)
 
-	rw.receiver.nextConsumer.ConsumeTraceData(ctx, *tracedata)
+	err := ocr.nextConsumer.ConsumeTraceData(ctx, *tracedata)
+	if err != nil {
+		observability.RecordMetricsForTraceReceiver(longLivedCtx, 0, len(tracedata.Spans))
+		span.Annotate([]trace.Attribute{
```

Comment:
It was already there before but I think that `AddAttribute` will be better since the time of an annotation is not specially relevant in this case and the span already has a timestamp (same below).

Reply:
Fixing.